core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -242,10 +241,179 @@ public void messageReceived(final ReplicaRequest request, final TransportChannel
         }
     }
 
+    class PrimaryOperationTransportHandler implements TransportRequestHandler<Request> {
+
+        @Override
+        public void messageReceived(Request request, final TransportChannel channel) throws Exception {
+            ActionListener<Response> actionListener = new ActionListener<Response>() {
+                @Override
+                public void onResponse(Response response) {
+                    try {
+                        channel.sendResponse(response);
+                    } catch (Throwable t) {
+                        onFailure(t);
+                    }
+                }
+
+                @Override
+                public void onFailure(Throwable e) {
+                    try {
+                        channel.sendResponse(e);
+                    } catch (Throwable t) {
+                        logger.warn("failed to send response for a write request", t);
+                    }
+                }
+            };
+            new AsyncPrimaryAction(request, actionListener).run();
+        }
+    }
+
+    class AsyncPrimaryAction extends AbstractRunnable {
+
+        private final ClusterStateObserver observer;
+        private final ActionListener<Response> listener;
+        private final InternalRequest internalRequest;
+        private final Request request;
+        private final AtomicBoolean finished = new AtomicBoolean(false);
+        private volatile Releasable indexShardReference;
+
+        public AsyncPrimaryAction(Request request, ActionListener<Response> listener) {
+            this.observer = new ClusterStateObserver(clusterService, request.timeout(), logger);
+            this.listener = listener;
+            this.request = request;
+            this.internalRequest = new InternalRequest(request);
+            this.internalRequest.concreteIndex(request.internalShardRouting.getIndex());
+        }
+
+        @Override
+        public void onFailure(Throwable t) {
+            finishAsFailed(t);
+        }
+
+        @Override
+        protected void doRun() throws Exception {
+            final ShardRouting primary = request.internalShardRouting;
+            // Although this gets executed locally, this more of an assertion, but if change the primary action
+            // to be performed remotely this check is important to check before performing the action:
+            if (clusterService.localNode().id().equals(primary.currentNodeId()) == false) {
```

Comment:
do we need to do this? if we fail to find the shard in the indicesService, we'll throw an exception anyhow?